### PR TITLE
Add config option for positive bijector + re-introduce positive lower bound

### PR DESCRIPTION
**Pull request by jahall**
Based on @awav's second TODO list item in #1144: _"Current positive transform uses softplus, but in some cases it is "preferable" to use exp instead; it would be nice to have an option in the gpflow.config, such that the user could choose global positive transformation"_

Still a work in progress but putting it up here for early feedback (as I calibrate coding style for this project).

### CODE REVIEW

**Comment by jahall** [comment]
Also re-introducing positive lower bound (set to `1e-10` as per @frgsimpson's comment in #1110 for now) and fixing associated bug.

File: gpflow/config/__config__.py [hunk 1]
```diff
@@ -23,7 +30,8 @@ class _Values(enum.Enum):
     will be `GPFLOW_SUMMARY_FMT`."""
     INT = np.int32
     FLOAT = np.float64
-    POSITIVE_MINIMUM = None
+    POSITIVE_BIJECTOR = "softplus"
+    POSITIVE_MINIMUM = 1e-10
```
Comment (awav) [inline gpflow/config/__config__.py]: I would recommend to keep same as before `1e-6`. @frgsimpson can change it to 1e-10, when he needs it :)

File: gpflow/config/__config__.py [hunk 2]
```diff
@@ -124,12 +139,18 @@ def set_default_jitter(value: float):
     set_config(replace(config(), jitter=value))
 
 
-def set_default_summary_fmt(value: str):
-    formats = tabulate.tabulate_formats + ['notebook', None]
-    if value not in formats:
-        raise ValueError(f"Summary does not support '{value}' format")
+_VALID_POSITIVE_BIJECTORS = {"exp", "softplus"}
```
Comment (awav) [inline gpflow/config/__config__.py]: Could you put it either into the function or in the beginning of the file. (I prefer function itself)

File: gpflow/utilities/bijectors.py [hunk 3]
```diff
@@ -5,17 +5,50 @@
...
-    lower_value = config.default_positive_minimum()
-    if lower_value is None:
-        return tfp.bijectors.Softplus()
+_POSITIVE_BIJECTOR_MAP = {
+    "exp": tfp.bijectors.Exp,
+    "softplus": tfp.bijectors.Softplus,
+}
+
+
+class Shift(tfp.bijectors.AffineScalar):
```
Comment (awav) [inline gpflow/utilities/bijectors.py]: * Check the comment in the test about printing.
* I believe we don't need this custom class

File: gpflow/utilities/bijectors.py [hunk 4]
```diff
@@ -5,17 +5,50 @@
...
+             bijector: Optional[str] = None) -> tfp.bijectors.Bijector:
+    """
+    Returns a positive bijector (a reversible transformation from real to positive numbers).
+
+    :param lower: lower bound override (defaults to config.default_positive_minimum())
+    :param bijector: bijector method override (defaults to config.default_positive_bijector())
+    :returns: a bijector instance
+    """
+    bijector = _get_base_positive_bijector(bijector)
+    if lower is None:
```
Comment (awav) [inline gpflow/utilities/bijectors.py]:
```suggestion
    default = config.default_positive_minimum()
    lower = default if lower is None else lower
```

Comment (awav) [reply gpflow/utilities/bijectors.py]: You need update two lines instead of one, github doesn't allow to do suggestions for multiple lines.

File: gpflow/utilities/bijectors.py [hunk 5]
```diff
@@ -5,17 +5,50 @@
...
+        lower = config.default_positive_minimum()
+    if lower is not None:
+        # Apply lower bound shift after applying base positive bijector
+        shift = Shift(shift=to_default_float(lower))
+        bijector = tfp.bijectors.Chain([bijector, shift])
```
Comment (awav) [inline gpflow/utilities/bijectors.py]:
```suggestion
        bijector = tfp.bijectors.Chain([shift, bijector])
```

Shift should be first

File: gpflow/utilities/bijectors.py [hunk 6]
```diff
@@ -5,17 +5,50 @@
...
-    bijectors = [tfp.bijectors.AffineScalar(shift=shift), tfp.bijectors.Softplus()]
-    return tfp.bijectors.Chain(bijectors)
+def _get_base_positive_bijector(bijector_name: Optional[str] = None):
+    if bijector_name is None:
+        bijector_name = config.default_positive_bijector()
+    return _POSITIVE_BIJECTOR_MAP[bijector_name]()
 
 
-triangular = tfp.bijectors.FillTriangular
+def triangular():
```
Comment (awav) [inline gpflow/utilities/bijectors.py]: Could you add type annotation for return value?

File: gpflow/utilities/bijectors.py [hunk 7]
```diff
@@ -5,17 +5,50 @@
...
+        # Apply lower bound shift after applying base positive bijector
+        shift = Shift(shift=to_default_float(lower))
+        bijector = tfp.bijectors.Chain([bijector, shift])
+    return bijector
+
 
-    shift = to_default_float(lower_value)
-    bijectors = [tfp.bijectors.AffineScalar(shift=shift), tfp.bijectors.Softplus()]
-    return tfp.bijectors.Chain(bijectors)
+def _get_base_positive_bijector(bijector_name: Optional[str] = None):
```
Comment (awav) [inline gpflow/utilities/bijectors.py]: Could you add type annotation to return value?

File: tests/test_printing.py [hunk 8]
```diff
@@ -186,32 +186,32 @@ def create_model():
...
------------------------------------------  ---------  -----------  -----------  -------  -------  -------\n\
-Product.kernels[0].kernels[0].variance     Parameter  Softplus     True         ()       float64        1\n\
-Product.kernels[0].kernels[0].lengthscale  Parameter  Softplus     False        ()       float64        2\n\
-Product.kernels[0].kernels[1].variance     Parameter  Softplus     True         ()       float64        1\n\
-Product.kernels[0].kernels[1].lengthscale  Parameter  Softplus     False        ()       float64        2\n\
-Product.kernels[1].kernels[0].variance     Parameter  Softplus     True         ()       float64        1\n\
-Product.kernels[1].kernels[0].lengthscale  Parameter  Softplus     False        ()       float64        2\n\
-Product.kernels[1].kernels[1].variance     Parameter  Softplus     True         ()       float64        1\n\
-Product.kernels[1].kernels[1].lengthscale  Parameter  Softplus     False        ()       float64        2"""
+name                                       class      transform         trainable    shape    dtype      value\n\
```
Comment (awav) [inline tests/test_printing.py]: I would recommend to set config lower value to None, this test printing is going to be pain in the neck in general. Let's just constrain it to limited settings.

File: gpflow/config/__config__.py (hunk 1 above)
Comment (frgsimpson) [reply gpflow/config/__config__.py]: @awav  I'd really advise against it. For example, think of the consequences of imposing the positive transform on likelihood.variance (which I guess is the default). A 1e-6 minimum means that no GPflow model can reach an absolute precision of better than 1e-3!

File: gpflow/utilities/bijectors.py (hunk 4 above)
Comment (jahall) [reply gpflow/utilities/bijectors.py]: I think the current code is cleaner. It also only fetches the default value when necessary rather than on every call. Any strong preference for this suggestion?

File: gpflow/utilities/bijectors.py (hunk 5 above)
Comment (jahall) [reply gpflow/utilities/bijectors.py]: Actually I think bijector needs to be first - I believe that was the cause of the original bug in #1110. Here is a counterexample (in some kind of pseudo-code...):
```
unconstrained = -20.0
constrained = softplus(unconstrained + 1e-6)
constrained > 1e-6  # False!
```
whereas
```
unconstrained = -20.0
constrained = softplus(unconstrained) + 1e-6
constrained > 1e-6  # True
```

File: tests/test_printing.py (hunk 8 above)
Comment (jahall) [reply tests/test_printing.py]: I'm happy to set these tests back to lower value = None. But I would like to add an explicit test in for the case where it is not None and keep the more informative print string:
```
transform
----------------
Softplus + Shift
```
instead of the uninformative
```
transform
---------
Chain
```

File: gpflow/utilities/bijectors.py (hunk 3 above)
Comment (jahall) [reply gpflow/utilities/bijectors.py]: I agree this custom class _might_ be overkill...but the benefit of an informative print string (especially as this will become the new default) is compelling.

File: tests/test_printing.py (hunk 8 above)
Comment (jahall) [reply tests/test_printing.py]: ...though maybe this is still overkill

File: gpflow/utilities/bijectors.py (hunk 3 above)
Comment (jahall) [reply gpflow/utilities/bijectors.py]: Resolving this conversation - can move discussion to `test_printing`

File: gpflow/config/__config__.py (hunk 1 above)
Comment (jahall) [reply gpflow/config/__config__.py]: I feel like erring on the side of having a lower default makes sense...but up to you guys.

File: gpflow/utilities/bijectors.py (hunk 4 above)
Comment (awav) [reply gpflow/utilities/bijectors.py]: :) my preference would be to switch to python 3.8 and use walrus operator. As it is not really possible, therefore don't have a strong opinion about it.

File: gpflow/utilities/bijectors.py (hunk 5 above)
Comment (awav) [reply gpflow/utilities/bijectors.py]: Indeed, I checked documentation, https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/Chain. It is clearer what's happening. Good catch!

File: gpflow/utilities/bijectors.py (hunk 4 above)
Comment (jahall) [reply gpflow/utilities/bijectors.py]: Yes! Bring on the walrus...one day

File: gpflow/utilities/bijectors.py (hunk 3 above)
Comment (awav) [inline gpflow/utilities/bijectors.py]: This class should be removed, it will confuse people. The print should reflect existing TFP bijectors, not GPflow's aliases.

Comment (jahall) [reply gpflow/utilities/bijectors.py]: Yes, this makes sense

File: gpflow/config/__config__.py (hunk 1 above)
Comment (awav) [reply gpflow/config/__config__.py]: Okay, I think we need to revert it back to the None value and think about the documentation for this in next PRs.

Comment (jahall) [reply gpflow/config/__config__.py]: At least we know it doesn't break anything now.

File: gpflow/config/__config__.py [hunk 9]
```diff
@@ -124,12 +139,17 @@ def set_default_jitter(value: float):
...
-def set_default_summary_fmt(value: str):
-    formats = tabulate.tabulate_formats + ['notebook', None]
-    if value not in formats:
-        raise ValueError(f"Summary does not support '{value}' format")
+def set_default_positive_bijector(value: str):
+    if not isinstance(value, str):
+        raise TypeError(f"Expected a string, but got a {type(value)}")
 
-    set_config(replace(config(), summary_fmt=value))
+    valid_bijectors = {"exp", "softplus"}
```
Comment (st--) [inline gpflow/config/__config__.py]: Could this simply be taken as `gpflow.utilities.bijectors._POSITIVE_BIJECTOR_MAP.keys()` instead? More DRY.

File: gpflow/config/__config__.py [hunk 10]
```diff
@@ -124,12 +139,17 @@ def set_default_jitter(value: float):
     set_config(replace(config(), jitter=value))
 
 
-def set_default_summary_fmt(value: str):
-    formats = tabulate.tabulate_formats + ['notebook', None]
-    if value not in formats:
-        raise ValueError(f"Summary does not support '{value}' format")
+def set_default_positive_bijector(value: str):
+    if not isinstance(value, str):
+        raise TypeError(f"Expected a string, but got a {type(value)}")
```
Comment (st--) [inline gpflow/config/__config__.py]: This error message could be a bit confusing - it doesn't actually expect "just a string", but a specific set of values. As a user, it would be more helpful to directly get the message "It has to be 'A' or 'B', not 5.0". Could deal with the lowering by "if isinstance(value, str): value = value.lower()"; if it's not a string it still won't be in the set of valid options.

Comment (st--) [reply gpflow/config/__config__.py]: (Similar to set_default_summary_fmt)

File: gpflow/config/__config__.py (hunk 9 above)
Comment (awav) [reply gpflow/config/__config__.py]: @st-- , @jahall 

Yes! But, it should be in the config itself as an immutable field. The other way around is wrong.

```python
@dataclass(frozen=True)
class Config:
    """
    Immutable object for storing global GPflow settings
    Args:
        int: Integer data type, int32 or int64.
        float: Float data type, float32 or float64
        jitter: Jitter value. Mainly used for for making badly conditioned matrices more stable.
            Default value is `1e-6`.
        positive_bijector: Method for positive bijector, either "softplus" or "exp".
            Default is "softplus".
        positive_minimum: Lower level for the positive transformation.
        summary_fmt: Summary format for module printing.
    """

    int: type = field(default_factory=lambda: default(_Values.INT))
    float: type = field(default_factory=lambda: default(_Values.FLOAT))
    jitter: float = field(default_factory=lambda: default(_Values.JITTER))
    positive_bijector: str = field(default_factory=lambda: default(_Values.POSITIVE_BIJECTOR))
    positive_minimum: float = field(default_factory=lambda: default(_Values.POSITIVE_MINIMUM))
    summary_fmt: str = field(default_factory=lambda: default(_Values.SUMMARY_FMT))
    positive_type: Dict[str, type] = dict(exp=tfp.bijectors.Exp, softplus=tfp.bijectors.Softplus)
```

The only thing I don't like is that dictionary itself here is immutable.

Comment (jahall) [reply gpflow/config/__config__.py]: Yep, agree that config shouldn't be importing from other modules. Happy to implement @awav's suggestion.

Comment (jahall) [reply gpflow/config/__config__.py]: I refactored so that the actual tfp.bijector.Bijector instance lives in `Config.positive_bijector` rather than a plain string so should resolve these issues...though maybe we don't want "complex" datastructures like these in the config?

File: gpflow/config/__config__.py [hunk 11]
```diff
@@ -143,10 +173,17 @@ def set_default_positive_minimum(value: float):
...
+        raise ValueError(f"Summary does not support '{value}' format")
+
+    set_config(replace(config(), summary_fmt=value))
+
+
 @contextlib.contextmanager
 def as_context(temporary_config: Optional[Config] = None):
-    '''Ensure that global configs defaults, with a context manager. Useful
-    for testing.'''
+    """Ensure that global configs defaults, with a context manager. Useful for testing."""
```
Comment (st--) [inline gpflow/config/__config__.py]: This 
```suggestion
    """Ensure that global configs defaults, with a context manager. Useful for testing."""!
```
still doesn't make sense to me :( not your code but wanted to flag it up now that I noticed if you don't mind figuring out the intended meaning...

File: gpflow/config/__config__.py [hunk 12]
```diff
@@ -47,13 +68,17 @@ class Config:
...
             Default value is `1e-6`.
+        positive_bijector: Method for positive bijector, either "softplus" or "exp".
+            Default is "softplus".
         positive_minimum: Lower level for the positive transformation.
         summary_fmt: Summary format for module printing.
     """
 
     int: type = field(default_factory=lambda: default(_Values.INT))
     float: type = field(default_factory=lambda: default(_Values.FLOAT))
     jitter: float = field(default_factory=lambda: default(_Values.JITTER))
```
Comment (st--) [inline gpflow/config/__config__.py]: I don't think this will ensure that the environment variables are converted to type objects / float values (and same for `positive_minimum` below)! again -not fault of this PR- but thought important to flag this up now that I spotted it

File: gpflow/utilities/bijectors.py [hunk 13]
```diff
@@ -5,17 +5,31 @@
...
+        shift = tfp.bijectors.AffineScalar(shift=to_default_float(lower))
+        bijector = tfp.bijectors.Chain([bijector, shift])
+    return bijector
+
+
+def triangular() -> tfp.bijectors.Bijector:
+    """
+    Returns instance of a triangular bijector.
+    """
+    return tfp.bijectors.FillTriangular()
```
Comment (st--) [inline gpflow/utilities/bijectors.py]: Is it wrong to re-use bijectors ?
In that case, the code in `__config__.py` needs changing, too, as there the config's `positive` bijector is already the instance, not the type (`default_positve_bijector()` returns the config value, doesn't call it).

File: gpflow/utilities/bijectors.py (hunk 5 above)
Comment (st--) [reply gpflow/utilities/bijectors.py]: It would be great if you could add this as a test!

File: gpflow/config/__config__.py (hunk 11 above)
Comment (awav) [reply gpflow/config/__config__.py]: Looks like I was drunk when were writing this :) @jahall skip it, I will fix it in a separate single PR with other doc fixes.

File: gpflow/utilities/bijectors.py (hunk 5 above)
Comment (awav) [reply gpflow/utilities/bijectors.py]: Test for what, that people read the documentation :) ?

File: gpflow/utilities/bijectors.py (hunk 13 above)
Comment (awav) [reply gpflow/utilities/bijectors.py]: Yes, it is wrong to re-use bijectors. They have cache inside.

File: gpflow/config/__config__.py (hunk 12 above)
Comment (awav) [reply gpflow/config/__config__.py]: Indeed, @jahall I will fix it in separate PR.

File: gpflow/config/__config__.py [hunk 14]
```diff
@@ -33,10 +42,22 @@ def name(self):
...
+def _to_bijector(value: str) -> tfp.bijectors.Bijector:
+    bijector_map = {
+        "exp": tfp.bijectors.Exp,
+        "softplus": tfp.bijectors.Softplus,
+    }
+    if isinstance(value, str):
+        value = value.lower()
+    if value not in bijector_map:
+        raise ValueError(f"`{value}` not in set of valid bijectors: {sorted(bijector_map)}")
+    return bijector_map[value]()
```
Comment (st--) [inline gpflow/config/__config__.py]: This already instantiates the bijector, so the config will always return the same object, which means sharing the bijector which according to @awav is actually wrong...

File: gpflow/utilities/bijectors.py (hunk 13 above)
Comment (st--) [reply gpflow/utilities/bijectors.py]: Then have to change the __config__ code, too.

File: gpflow/config/__config__.py (hunk 14 above)
Comment (jahall) [reply gpflow/config/__config__.py]: Good point - I think having the actual bijectors in the config is maybe too complex - let me refactor.

Comment (st--) [reply gpflow/config/__config__.py]: It should be fine to have the definitions in `bijectors.py` and still get the valid keys from there inside `__config__.py` without having to manually duplicate! :)

Comment (st--) [reply gpflow/config/__config__.py]:
```suggestion
    return bijector_map[value]
```
and then instantiate it inside bijectors.py's `positive()` :)

File: gpflow/utilities/bijectors.py (hunk 5 above)
Comment (jahall) [reply gpflow/utilities/bijectors.py]: Soooo, turns out I was totally wrong. `Chain` applies the transformations in reverse order so the original implementation was correct - good job you suggested adding a test!

File: tests/test_utilities.py [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        bijector = positive(base=override_bijector)
+        assert isinstance(bijector, expected_class)
+
+
+def test_positive_calculation_order():
+    value, lower = -10.0, 10.0
+    expected = np.exp(value) + lower
+    with as_context(Config(positive_bijector="exp", positive_minimum=lower)):
+        result = positive()(value).numpy()
+    assert np.isclose(result, expected)
```
Comment (st--) [inline tests/test_utilities.py]:
```suggestion
    assert np.isclose(result, expected)
    assert result > lower
```
